Add ENSIP: Contract Self-Naming

nxt3d opened the pull request:
## Summary

This ENSIP extends ENSIP-19 with two interfaces that let contracts control their own reverse names:

- `IContractName` (selector `0x75d0c0dc`): the contract publishes its own ENS name through `contractName()` and emits a `ContractNameSet` event when the name changes. Indexers and resolvers that support this interface can discover and follow the contract's name directly. No additional reverse-resolver registration is required.
- `IContractNamer` (selector `0x6f3ff726`): the contract delegates naming authority to one or more addresses through `isContractNamer(address)`. This generalizes the previous Ownable-only delegation pattern so multisigs, role-based access controls, and custom authorization logic can serve as the naming authority.

Both interfaces are defined in the post-audit branch of [ensdomains/contracts-v2](https://github.com/ensdomains/contracts-v2/tree/post-audit/contracts/src/reverse-registrar/interfaces).

## Precedence

When a value is registered in the reverse resolver registry and `contractName()` returns a value for the same contract, the reverse resolver registry takes precedence.

## ERC-165

ERC-165 advertisement is RECOMMENDED for both interfaces. Reverse registrars MUST detect `IContractNamer` through ERC-165 because of the security implications of selector collisions in the authorization path.

## Backwards Compatibility

Fully backwards compatible with ENSIP-19. Contracts that do not implement either interface continue to use existing reverse-registration flows.

## History

This PR replaces the prior draft, which used ERC-8042 Diamond Storage for a self-declared namehash. The new approach uses the interfaces published in `ensdomains/contracts-v2` post-audit, which were not yet defined when the original draft was written.

nxt3d committed:
Add ENSIP: Contract Self-Naming via ERC-8049

This ENSIP enables contracts to declare their own reverse names using ERC-8049 metadata with Optional Diamond Storage extension. Contracts can self-declare their ENS name during deployment, and any account can then trustlessly verify and permissionlessly register the contract's reverse name in the registrar.

Key features:
- Contracts declare reverse names via ERC-8049 metadata at key 'eth.ens.reverse-name'
- Predictable storage locations enable trustless verification
- Permissionless registration removes burden from contract deployers
- Compatible with ENSIP-19
- Includes verified storage slot calculations

nxt3d committed:
Add ENSIP: Contract Self-Naming via ERC-8049

This ENSIP enables contracts to declare their own reverse names using ERC-8049 metadata with Optional Diamond Storage extension. Contracts can self-declare their ENS name during deployment, and any account can then trustlessly verify and permissionlessly register the contract's reverse name in the registrar.

Key features:
- Contracts declare reverse names via ERC-8049 metadata at key 'eth.ens.reverse-name'
- Predictable storage locations enable trustless verification
- Permissionless registration removes burden from contract deployers
- Compatible with ENSIP-19
- Includes verified storage slot calculations

nxt3d committed:
Clarify Motivation section with alternative approaches

- Note that contract owner may not always exist
- Mention alternative of making custom external call in constructor

nxt3d committed:
Clarify Motivation section with alternative approaches

- Note that contract owner may not always exist
- Mention alternative of making custom external call in constructor

nxt3d committed:
Merge branch 'ensdomains:master' into contract-self-naming

nxt3d committed:
Fix terminology and clarify rationale in contract-self-naming.md

- Change 'ABI-encoded' to 'stored as bytes' for value format (correct terminology)
- Clarify single-slot verification applies to names under 32 characters
- Update rationale to emphasize admin-free contracts use case

nxt3d committed:
Fix terminology and clarify rationale in contract-self-naming.md

- Change 'ABI-encoded' to 'stored as bytes' for value format (correct terminology)
- Clarify single-slot verification applies to names under 32 characters
- Update rationale to emphasize admin-free contracts use case

nxt3d committed:
Update title to 'Contract Self-Naming' (remove 'via ERC-8049')

nxt3d committed:
Update title to 'Contract Self-Naming' (remove 'via ERC-8049')

nxt3d committed:
Update ENSIP number format to ENSIP-X (number not yet assigned)

nxt3d committed:
Update ENSIP number format to ENSIP-X (number not yet assigned)

nxt3d committed:
ENSIP: switch contract self-naming to ERC-8042 diamond storage

This updates the Contract Self-Naming ENSIP to remove the ERC-8049 dependency and instead use
an ERC-8042 diamond storage slot keyed by "eth.ens.reverse-name".

Key changes:
- Replace the ERC-8049 Optional Diamond Storage metadata mapping approach with direct ERC-8042
  storage location addressing.
- Define the storage location as keccak256("eth.ens.reverse-name"), matching ERC-8042’s formula
  (erc8042(id) = keccak256(id)).
- Specify that the value stored at this slot is a bytes32 namehash (rather than a UTF-8 bytes
  string), keeping the declaration single-slot and efficient to verify.
- Update the Solidity example to the canonical diamond-storage struct pattern:
  - STORAGE_POSITION constant
  - getStorage() that assigns s.slot via assembly
  - @custom:storage-location erc8042:eth.ens.reverse-name annotation
  - set ReverseNameStorage.reverseNameHash during deployment
- Update abstract/motivation/rationale/backwards compatibility wording and references to reflect
  the simplified ERC-8042-only approach.
- Add raffy.eth as a contributor.

nxt3d committed:
ENSIP: switch contract self-naming to ERC-8042 diamond storage

This updates the Contract Self-Naming ENSIP to remove the ERC-8049 dependency and instead use
an ERC-8042 diamond storage slot keyed by "eth.ens.reverse-name".

Key changes:
- Replace the ERC-8049 Optional Diamond Storage metadata mapping approach with direct ERC-8042
  storage location addressing.
- Define the storage location as keccak256("eth.ens.reverse-name"), matching ERC-8042’s formula
  (erc8042(id) = keccak256(id)).
- Specify that the value stored at this slot is a bytes32 namehash (rather than a UTF-8 bytes
  string), keeping the declaration single-slot and efficient to verify.
- Update the Solidity example to the canonical diamond-storage struct pattern:
  - STORAGE_POSITION constant
  - getStorage() that assigns s.slot via assembly
  - @custom:storage-location erc8042:eth.ens.reverse-name annotation
  - set ReverseNameStorage.reverseNameHash during deployment
- Update abstract/motivation/rationale/backwards compatibility wording and references to reflect
  the simplified ERC-8042-only approach.
- Add raffy.eth as a contributor.

nxt3d committed:
ENSIP: clarify namehash() usage in ERC-8042 example

Update the example contract in the Contract Self-Naming ENSIP to explicitly use namehash("mycontract.eth").

nxt3d committed:
ENSIP: clarify namehash() usage in ERC-8042 example

Update the example contract in the Contract Self-Naming ENSIP to explicitly use namehash("mycontract.eth").

nxt3d committed:
ENSIP: minor example/rationale cleanup

- Rename ReverseNameStorage field to match the intended namehash wording in the example.
- Clarify in the rationale that storing the namehash is more efficient than storing the plain string.

clowestab commented on ensips/contract-self-naming.md:
```diff
@@ -0,0 +1,91 @@
+
+## Abstract
+
+This ENSIP extends ENSIP-19 to enable contracts to declare their own reverse names using ERC-8042 Diamond Storage. By storing a namehash of the reverse name in a known storage location, contracts can self-declare their ENS name. Any account can then register this name using the reverse registrar on L1, or using an L2 reverse registrar. This enables trustless, permissionless reverse name registrat ...
```
You mean the namehash of the primary name?

_"Any account can then register this name using the reverse registrar on L1, or using an L2 reverse registrar."_

This spec doesn't outline **how**..?

Arachnid commented:
Wouldn't a simpler implementation be to define an interface with a predicate function that allows the contract to explicitly accept or reject a primary name?

Also, presumably the idea here is to update the reverse registrar to support checking this and setting the name accordingly, but this isn't mentioned anywhere.

nxt3d commented:
> Wouldn't a simpler implementation be to define an interface with a predicate function that allows the contract to explicitly accept or reject a primary name?
>


By predicate function, do you mean something like `isReverseName(bytes32 namehash) returns (bool)`? What benefit would this have over saving the namehash as a named storage slot value?

With a named storage slot using ERC‑8042 Diamond Storage, the storage value itself acts as the verification. If the `namehash` exists at that slot and matches, it is valid. This avoids function specification overhead, is chain‑agnostic because there is no need to know the reverse registrar address, uses minimal gas via a single `sstore`, and is upgrade‑safe since storage persists across proxy upgrades.

**Example:**

```solidity
constructor() {
    // Declare ENS reverse name using ERC-8042 Diamond Storage
    // Slot: keccak256("eth.ens.reverse-name")
    // Value: namehash("mycontract.eth")
    // This allows any account to permissionlessly register the reverse name
    assembly {
        sstore(
            0x09ded414ae6c0ce389342caf0619071d5be1687a6f7314e74bcc7cfa1a0df4bf,
            0x6ff093714e95f3f123db7d14a81db07d48e7e0f7ca7ef35ca446facd9d5e7c1c
        )
    }
}
```

> Also, presumably the idea here is to update the reverse registrar to support checking this and setting the name accordingly, but this isn't mentioned anywhere.

I agree, I can work on that and document it once the self-naming mechanism is finalized.

adraffy commented:
How about:

On L1, a contract can name itself by implementing `INameResolver`.  The detection method can be ERC-165.  The `addr.reverse` resolver can implement this check.  This dodges the ownership requirement. 

On L2 (or any contract that interfaces with the crosschain reverse registrars), since we want names stored in one contract for read simplicity (so the name needs copied into the registrar), the only option currently is to implement `owner()`.

However, a similar mechanism like above would work:  there can be a function on the registrar that anyone can call, and if the address supports `INameResolver`, it copies that name accordingly.

In both situations, the requirement would be the `name()` response must be onchain.

Instead of ERC-165, it could just be a blind call with a non-empty string response.

It would be up to the contract to check if `node` matches.  Alternatively, we could use a different function (maybe that's what was meant by predicate function above.)

Arachnid commented:
> This avoids function specification overhead, 

Except it requires referencing ERC-8042, which is large and overengineered.

> is chain‑agnostic because there is no need to know the reverse registrar address

Nor does my suggested alternative require this

> uses minimal gas via a single sstore

20k gas - while a simple constant-returning function uses much less

> and is upgrade‑safe since storage persists across proxy upgrades.

Which is a cost, rather than a benefit, since you may want to change the primary name

nxt3d commented:
> Except it requires referencing ERC-8042, which is large and overengineered.

We can remove the dependency on ERC-8042 and simply have the storage location use the same scheme, which I think is simpler than ERC-7201?

> Nor does my suggested alternative require this.

I agree.

> 20k gas, while a simple constant-returning function uses much less.

Right, if it is a constant, I agree.

> Which is a cost, rather than a benefit, since you may want to change the primary name.

How does this affect whether it can be updated or not? If we use a function, then that friction is required for every upgrade, whether or not the developer wants that friction, because the developer needs to ensure that the function call persists across upgrades.

I also think this may be a key part of the design: 
> This avoids function specification overhead.

We might also decide in the future to allow L2 reverse names to be registered on L1, in which case we could use an inclusion proof to verify the contract name.

---------------------------------------------------

Another idea would be to just do both methods. If the goal is to make it as easy for contract developers as possible to set their reverse name, then we could offer both methods as options, allowing for developers to decide what works best for them.

Arachnid commented:
> How does this affect whether it can be updated or not? If we use a function, then that friction is required for every upgrade, whether or not the developer wants that friction, because the developer needs to ensure that the function call persists across upgrades.

How is leaving a piece of code unmodified 'friction'?

nxt3d commented:
> > How does this affect whether it can be updated or not? If we use a function, then that friction is required for every upgrade, whether or not the developer wants that friction, because the developer needs to ensure that the function call persists across upgrades.
> 
> How is leaving a piece of code unmodified 'friction'?

Because it's necessary to continue to manage the interface, such as isReverseName(bytes32 namehash) returns (bool), across multiple contract upgrades, whereas setting the named storage value in the constructor can be done only once and persists across upgrades.

My original intention for this proposal was to try to make it as easy as possible for a contract to name itself. I couldn't think of anything easier than setting a storage value.

Arachnid commented:
> Because it's necessary to continue to manage the interface, such as isReverseName(bytes32 namehash) returns (bool), across multiple contract upgrades, whereas setting the named storage value in the constructor can be done only once and persists across upgrades.

This is a real stretch. I think having invisible state that *isn't* encompassed by the current version of the source is much more of a maintenance and transparency issue than a 3-line function you simply leave alone.

nxt3d commented:
> This is a real stretch. I think having invisible state that isn't encompassed by the current version of the source is much more of a maintenance and transparency issue than a 3-line function you simply leave alone.

I think that most implementers would like the value to be mutable, and therefore would create a setter function.

```setENSName(bytes32 namehash)```

I am happy to move forward with either design, or both. If I change this PR to just support `isReverseName(bytes32 namehash) returns (bool)`, is this something that we could move forward on? I could also add other methods such as ERC-1271 and `DEFAULT_ADMIN_ROLE` for access control.

nxt3d committed:
ENSIP: rewrite around IContractName + IContractNamer interfaces

Replaces the ERC-8042 Diamond Storage approach with two interfaces
published in ensdomains/contracts-v2 post-audit:

- IContractName (0x75d0c0dc): contract publishes its own name via
  contractName() and emits ContractNameSet on set/change. Self
  contained. No additional reverse-resolver registration required.
- IContractNamer (0x6f3ff726): contract delegates naming authority
  to one or more addresses via an isContractNamer(addr) predicate,
  generalizing the previous Ownable-only delegation pattern.

ERC-165 advertisement is RECOMMENDED for both interfaces. Reverse
registrars MUST detect IContractNamer through ERC-165 because of
the security implications of selector collisions in the
authorization path.

Adds a Precedence rule: when both a value is registered in the
reverse resolver registry and contractName() returns a value for
the same contract, the reverse resolver registry takes precedence.

Reframes Motivation as added options for contract authors.
